org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/FormatterManager.java

```diff
@@ -0,0 +1,198 @@
+
+	/**
+	 * Identifiers for the XML file.
+	 */
+	private final static String XML_NODE_ROOT= "profiles"; //$NON-NLS-1$
+	private final static String XML_NODE_PROFILE= "profile"; //$NON-NLS-1$
+	private final static String XML_NODE_SETTING= "setting"; //$NON-NLS-1$
+
+	private final static String XML_ATTRIBUTE_ID= "id"; //$NON-NLS-1$
+	private final static String XML_ATTRIBUTE_NAME= "name"; //$NON-NLS-1$
+	private final static String XML_ATTRIBUTE_PROFILE_KIND= "kind"; //$NON-NLS-1$
+	private final static String XML_ATTRIBUTE_VALUE= "value"; //$NON-NLS-1$
+
+	public FormatterManager() {
+	}
+
+	/**
+	 * Read the available profiles from the internal XML file and return them as
+	 * collection or <code>null</code> if the file is not a profile file.
+	 *
+	 * @param file
+	 *            The file to read from
+	 * @return returns a map of formatter options or <code>null</code>
+	 * @throws CoreException
+	 */
+	public Map<String, String> readSettingsFromFile(File file, String profileName) throws CoreException {
+		try (FileInputStream reader = new FileInputStream(file)) {
+			return readSettingsFromStream(new InputSource(reader), profileName);
+		} catch (IOException e) {
+			throw new CoreException(new Status(IStatus.WARNING, IConstants.PLUGIN_ID, e.getMessage(), e));
+		}
+	}
+
+	/**
+	 * Load profiles from a XML stream and add them to a map or <code>null</code> if the source is not a profile store.
+	 * @param inputSource The input stream
+	 * @return returns a list of <code>CustomProfile</code> or <code>null</code>
+	 * @throws CoreException
+	 */
+	public static Map<String, String> readSettingsFromStream(InputSource inputSource, String profileName) throws CoreException {
+
+		final ProfileDefaultHandler handler = new ProfileDefaultHandler(profileName);
+		try {
+			final SAXParserFactory factory = SAXParserFactory.newInstance();
+			final SAXParser parser = factory.newSAXParser();
+			parser.parse(inputSource, handler);
+		} catch (Exception e) {
+			throw new CoreException(new Status(IStatus.WARNING, IConstants.PLUGIN_ID, e.getMessage(), e));
+		}
+		return handler.getSettings();
+	}
+
+	public static void configureFormatter(PreferenceManager preferenceManager, ProjectsManager projectsManager) {
+		String formatterUrl = preferenceManager.getPreferences().getFormatterUrl();
+		Map<String, String> options = null;
+		if (formatterUrl != null) {
+			URL url = projectsManager.getUrl(formatterUrl);
+			if (url != null) {
+				try (InputStream inputStream = url.openStream()) {
+					InputSource inputSource = new InputSource(inputStream);
+					String profileName = preferenceManager.getPreferences().getFormatterProfileName();
+					options = FormatterManager.readSettingsFromStream(inputSource, profileName);
+				} catch (Exception e) {
+					JavaLanguageServerPlugin.logException(e.getMessage(), e);
+				}
+			} else {
+				JavaLanguageServerPlugin.logInfo("Invalid formatter:" + formatterUrl);
+			}
+		}
+		if (options != null && !options.isEmpty()) {
```

Comment:
The problem here is if new options is a subset of previous options, then some of those previous options will be reused. Instead, we should always merge the new options with the default (not previous) options.